SSA/Experimental/IntrinsicAsymptotics.lean

```diff
@@ -9,83 +9,133 @@ import Mathlib.Tactic.Ring
 
 open Ctxt (Var VarSet)
 
-/- 
-  # Datastructures 
+/-
+  # Datastructures
 -/
 
+inductive Op :  Type
+  | add : Op
+  | beq : Op
+  | cst : ℕ → Op
+  deriving DecidableEq
+
+def Op.sig : Op → List Ty
+  | .add => [.nat, .nat]
+  | .beq => [.nat, .nat]
+  | .cst _ => []
+
+def Op.outTy : Op → Ty
+  | .add => .nat
+  | .beq => .bool
+  | .cst _ => .nat
+
+inductive HVector (A : Ty → Type*) : List Ty → Type _
+  | nil : HVector A []
+  | cons {t : Ty} {l : List Ty} : A t → HVector A l → HVector A (t::l)
+
 /-- A very simple intrinsically typed expression. -/
-inductive IExpr : Ctxt → Ty → Type
-  | add (a b : Γ.Var .nat) : IExpr Γ .nat
-  /-- Nat literals. -/
-  | cst (n : Nat) : IExpr Γ .nat
+structure IExpr (Γ : Ctxt) (ty : Ty) : Type :=
+  (op : Op)
+  (ty_eq : ty = op.outTy)
+  (args : HVector (Ctxt.Var Γ) op.sig)
+
+def IExpr.cst {Γ : Ctxt} (n : ℕ) : IExpr Γ .nat  :=
+  { op := Op.cst n
+    ty_eq := rfl
+    args := .nil }
+
+def IExpr.add {Γ : Ctxt} (e₁ e₂ : Var Γ .nat) : IExpr Γ .nat :=
+  { op := Op.add
+    ty_eq := rfl
+    args := .cons e₁ <| .cons e₂ .nil }
 
 /-- A very simple intrinsically typed program: a sequence of let bindings. -/
 inductive ICom : Ctxt → Ty → Type where
   | ret (v : Γ.Var t) : ICom Γ t
   | lete (e : IExpr Γ α) (body : ICom (Γ.snoc α) β) : ICom Γ β
 
-/-- `Lets Γ₁ Γ₂` is a sequence of lets which are well-formed under context `Γ₂` and result in 
+/-- `Lets Γ₁ Γ₂` is a sequence of lets which are well-formed under context `Γ₂` and result in
     context `Γ₁`-/
 inductive Lets : Ctxt → Ctxt → Type where
   | nil {Γ : Ctxt} : Lets Γ Γ
   | lete (body : Lets Γ₁ Γ₂) (e : IExpr Γ₂ t) : Lets Γ₁ (Γ₂.snoc t)
 
 
 
-/- 
+/-
   # Definitions
 -/
 
-def IExpr.denote : IExpr Γ ty → (Γv : Γ.Valuation) → ty.toType
+def HVector.map {A B : Ty → Type*} (f : ∀ (t : Ty), A t → B t) :
+    ∀ {l : List Ty}, HVector A l → HVector B l
+  | [], .nil => .nil
+  | t::_, .cons a as => .cons (f t a) (map f as)
+
+def HVector.foldl {A : Ty → Type*} {B : Type*} (f : ∀ (t : Ty), B → A t → B) :
+    ∀ {l : List Ty}, B → HVector A l → B
+  | [], b, .nil => b
+  | t::_, b, .cons a as => foldl f (f t b a) as
+
+@[reducible]
+def Op.denote : (op : Op) →
+    HVector Ty.toType op.sig → op.outTy.toType
   | .cst n, _ => n
-  | .add a b, Γv => Γv a + Γv b
+  | .add, .cons a (.cons b .nil) => a + b
```

Comment:
that's pretty cool that we can just pattern match on these like that! Do you think it makes sense to add syntatic sugar for the pattern matching, so we don't have to call these constructors explicitly and it just sort of "looks like a tuple/list"?